### ISSUE TITLE
caching input features

### ISSUE THREAD

**Issue by amhajavi** [s3prl/s3prl#134]
Is there a way use cached extracted features in pretrain stage? For example extract spectrogram features once from the dataset and use it as input? The GPU server I am using does not have a very powerful CPU so the spectrogram extraction is a pain and doing it for 1M steps takes forever.

**Comment by andi611** [comment]
I believe this issue is highly related to #191, if #191 is resolved then this issue will be automatically resolved too. Since "cached extracted features in the pre-train stage" is essentially caching the "baseline" features.

**Comment by andi611** [comment]
> Does the TERA pretraining use CPU to extract spectrogram? Or GPU?

TERA uses the [preprocessor](https://github.com/s3prl/s3prl/blob/master/s3prl/upstream/baseline/preprocessor.py) you previously implemented, which I believe is on the CPU.

**Comment by leo19941227** [comment]
Oh actually the preprocessor is compatible with both CPU/GPU. I checked the TERA `pretrained_expert`, it gets the extracted spectrogram from dataloader instead of just waveform, so yes it is extracted by CPU.

However, I think this issue will not be automatically resolved as long with @191. The solution (if there is one) for SUPERB will be highly entangled with not only `upstream` but also `run_downstream.py`, `downstream/runner.py` and the `dataloader` of each downstream task, since loading the cached features requires huge amount of IO and is not possible to do it in `upstream` solely which will be extremely slow. So I think the code fixing @191 will not be directly reusable for the pretraining part.

**Comment by andi611** [comment]
Just some more information, it may be helpful in the future:

The huggingface datasets also use cache files,
once the preprocessing on a dataset is computed, it will automatically store it and reuse it next time you call the same preprocessing and dataset (can be reused on a separate execution but has to be on the same machine), in my recent experience in using it, it is very fast and efficient.

Maybe we can look into how they achieve this, the caching is self-contained in the datasets class, disentangled from other code. 

For more info, see [cache files](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=load_from#datasets.Dataset.cache_files) and the [cast](https://huggingface.co/docs/datasets/package_reference/main_classes.html?highlight=load_from#datasets.Dataset.cast) method (with the `load_from_cache_file` parameter).

**Comment by leo19941227** [comment]
Cool! We can dive into this more after we are more available xd
I believe we are both very occupied recently.

I am not sure how much the huggingface's dataset trick will help the IO bottleneck. I guess it does save time from the heavy CPU preprocessing without re-doing it. But one difficulty I am thinking is that the `all layers of hidden states`, `long sequence of audio data`, and `high dimensional hidden state` natures can make an already very large dataset many times larger and turn out even loading a single cached preprocessed utterance being very slow. I remember one year before when we are still pretraining with the target 1024-dim linear spectrogram for Mockingjay, even loading a single layer of pre-extracted 1024 spectrogram cause huge burden for IO and is much slower than just compute the linear spectrogram on-the-fly (either in CPU or accelerate by GPU). So I think the solution will need to involve some kind of data compression. I am not sure whether huggingface's caching do the compression or not, will need to investigate more. Or maybe there are much better storing file formats which can achieve much higher IO speed and we don't need to compress. Do you know there is any special design on loading the cached features for faster IO in huggingface? Or perhaps we can investigate this together later.

Beside caching features as the solution, I think another much more easy and feasible solution for this issue might be to put the spectrogram extraction in the downstream expert's forward and accelerated by GPU. OnlinePreprocessor's spectrogram extraction can be highly accelerated by GPU and no need to be done at dataloader. For the other preprocessing including spectrogram masking and attention mask preparation, I will guess that they are not computational bottleneck. As far as I know the wav2vec2 pretraining in fairseq is also doing SpecAugment on representations in the model forward instead of dataloader.

**Comment by leo19941227** [comment]
We are working on this as a new feature and will use Arrow for the underlying caching system. However, the current old codebase will not be supported since it is intrinsically hard to do it in a maintainable way. Hence I am closing this issue. This feature will be supported in a follow up new version
